core/src/providers/textsynth.rs

```diff
@@ -0,0 +1,485 @@
+            body["presence_penalty"] = json!(presence_penalty.unwrap());
+        }
+        if repetition_penalty.is_some() {
+            body["repetition_penalty"] = json!(repetition_penalty.unwrap());
+        }
+        if typical_p.is_some() {
+            body["typical_p"] = json!(typical_p.unwrap());
+        }
+
+        let req = Request::builder()
+            .method(Method::POST)
+            .uri(self.uri()?)
+            .header("Content-Type", "application/json")
+            .header(
+                "Authorization",
+                format!("Bearer {}", self.api_key.clone().unwrap()),
+            )
+            .body(Body::from(body.to_string()))?;
+
+        let res = cli.request(req).await?;
+        let status = res.status();
+        let body = hyper::body::aggregate(res).await?;
+        let mut b: Vec<u8> = vec![];
+        body.reader().read_to_end(&mut b)?;
+        let c: &[u8] = &b;
+
+        let response = match status {
+            hyper::StatusCode::OK => {
+                // print content of c
+                println!("RESPONSE: {}", String::from_utf8_lossy(c));
+                let completion: Completion = serde_json::from_slice(c)?;
+                Ok(completion)
+            }
+            hyper::StatusCode::TOO_MANY_REQUESTS => {
+                let error: Error = serde_json::from_slice(c).unwrap_or(Error {
+                    error: "Too many requests".to_string(),
+                });
+                Err(ModelError {
+                    message: format!("TextSynthAPIError: {}", error.error),
+                    retryable: Some(ModelErrorRetryOptions {
+                        sleep: Duration::from_millis(2000),
+                        factor: 2,
+                        retries: 8,
+                    }),
+                })
+            }
+            hyper::StatusCode::BAD_REQUEST => {
+                let error: Error = serde_json::from_slice(c).unwrap_or(Error {
+                    error: "Unknown error".to_string(),
+                });
+                Err(ModelError {
+                    message: format!("TextSynthAPIError: {}", error.error),
+                    retryable: None,
+                })
+            }
+            _ => {
+                let error: Error = serde_json::from_slice(c)?;
+                Err(ModelError {
+                    message: format!("TextSynthAPIError: {}", error.error),
+                    retryable: None,
+                })
+            }
+        }?;
+        Ok(response)
+    }
+}
+
+#[async_trait]
+impl LLM for TextSynthLLM {
+    fn id(&self) -> String {
+        self.id.clone()
+    }
+
+    async fn initialize(&mut self, credentials: Credentials) -> Result<()> {
+        match credentials.get("TEXTSYNTH_API_KEY") {
+            Some(api_key) => {
+                self.api_key = Some(api_key.clone());
+            }
+            None => match tokio::task::spawn_blocking(|| std::env::var("TEXTSYNTH_API_KEY")).await?
```

Comment:
Does this need spawn blocking ? Would have expected this to not be cpu bound

Reply:
yeah this is cargo-culted... we can clean-up everywhere. The thing is that std::env::var is kinda blocking but we can probably optimize